packages/monitor-v2/src/llm-bot/OptimisticOracleV2Client.ts

```diff
@@ -0,0 +1,113 @@
+    _provider: Provider,
+    _requests: Map<string, OptimisticOracleRequest> = new Map<string, OptimisticOracleRequest>(),
+    _fetchedBlockRanges?: BlockRange[]
+  ) {
+    super(_provider, _requests, _fetchedBlockRanges);
+  }
+
+  async getEventsWithPagination<E extends Event>(
+    filter: EventFilter,
+    fromBlock: number,
+    toBlock: number
+  ): Promise<E[]> {
+    const ooV2Contract = await getContractInstanceWithProvider<OptimisticOracleV2Ethers>(
+      "OptimisticOracleV2",
+      this.provider
+    );
+
+    const chainId = await this.provider.getNetwork().then((network) => network.chainId);
+
+    const maxBlockLookBack =
+      Number(process.env.MAX_BLOCK_LOOKBACK) ||
+      blockDefaults[chainId.toString() as keyof typeof blockDefaults]?.maxBlockLookBack ||
+      blockDefaults.other.maxBlockLookBack;
+
+    const searchConfig = {
+      fromBlock: fromBlock,
+      toBlock: toBlock,
+      maxBlockLookBack: maxBlockLookBack,
+    };
+
+    return paginatedEventQuery<E>(ooV2Contract, filter, searchConfig);
+  }
+
+  async applyRequestPriceEvent(requestPriceEvent: RequestPriceEvent): Promise<void> {
+    const body = tryHexToUtf8String(requestPriceEvent.args.ancillaryData);
+    const identifier = ethers.utils.parseBytes32String(requestPriceEvent.args.identifier);
+    const timestamp = requestPriceEvent.args.timestamp.toNumber();
+    const requester = requestPriceEvent.args.requester;
+    const requestId = calculateRequestId(body, identifier, timestamp, requester);
+
+    if (this.requests.has(requestId)) {
+      return;
+    }
+    const ooV2Contract = await getContractInstanceWithProvider<OptimisticOracleV2Ethers>(
+      "OptimisticOracleV2",
+      this.provider
+    );
+    const newRequest = new OptimisticOracleRequest({
+      requester: requestPriceEvent.args.requester,
+      identifier: ethers.utils.parseBytes32String(requestPriceEvent.args.identifier),
+      timestamp: requestPriceEvent.args.timestamp.toNumber(),
+      requestTx: requestPriceEvent.transactionHash,
+      type: OptimisticOracleType.PriceRequest,
+      body: tryHexToUtf8String(requestPriceEvent.args.ancillaryData),
+      blockNumber: requestPriceEvent.blockNumber,
+      transactionIndex: requestPriceEvent.transactionIndex,
+      isEventBased: await ooV2Contract
+        .getRequest(
+          requestPriceEvent.args.requester,
+          requestPriceEvent.args.identifier,
+          requestPriceEvent.args.timestamp,
+          requestPriceEvent.args.ancillaryData
+        )
+        .then((r) => r[4][0]),
+    });
+    this.requests.set(requestId, newRequest);
+  }
+
+  protected async updateOracleRequests(newRanges: BlockRange[]): Promise<void> {
+    const newRange = newRanges[newRanges.length - 1];
+
+    const ooV2Contract = await getContractInstanceWithProvider<OptimisticOracleV2Ethers>(
+      "OptimisticOracleV2",
+      this.provider
+    );
+
+    const requestPriceEvents = await this.getEventsWithPagination<RequestPriceEvent>(
```

Comment:
Only processing `RequestPrice` for now. The rest of events will go in a follow up PR.